Common/Securities/Option/QLOptionPriceModel.cs

```diff
@@ -0,0 +1,227 @@
+                        return Tuple.Create((decimal)option.delta(), (decimal)option.gamma());
+                    }
+                    catch (Exception)
+                    {
+                        if (optionSecurity.EnableGreekApproximation)
+                        {
+                            var step = 0.01;
+                            var initial = underlyingQuoteValue.value();
+                            underlyingQuoteValue.setValue(initial - step);
+                            var npvMinus = option.NPV();
+                            underlyingQuoteValue.setValue(initial + step);
+                            var npvPlus = option.NPV();
+                            underlyingQuoteValue.setValue(initial);
+
+                            return Tuple.Create((decimal)((npvPlus - npvMinus) / (2 * step)),
+                                                (decimal)((npvPlus - 2 * npv + npvMinus) / (step * step)));
+                        }
+                        else
+                            return Tuple.Create(0.0m, 0.0m);
+                    }
+                };
+
+                Func<double> reevalVega = () =>
+                {
+                    var step = 0.001;
+                    var initial = underlyingVolValue.value();
+                    underlyingVolValue.setValue(initial + step);
+                    var npvPlus = option.NPV();
+                    underlyingVolValue.setValue(initial);
+
+                    return (npvPlus - npv) / step;
+                };
+
+                Func<double> reevalTheta = () =>
+                {
+                    var step = 1.0 / 365.0;
+
+                    Settings.setEvaluationDate(settlementDate.AddDays(-1));
+                    var npvMinus = option.NPV();
+                    Settings.setEvaluationDate(settlementDate);
+
+                    return (npv - npvMinus) / step;
+                };
+
+                Func<double> reevalRho = () =>
+                {
+                    var step = 0.001;
+                    var initial = riskFreeRateValue.value();
+                    riskFreeRateValue.setValue(initial + step);
+                    var npvPlus = option.NPV();
+                    riskFreeRateValue.setValue(initial);
+
+                    return (npvPlus - npv) / step;
+                };
+
+                // producing output with lazy calculations of IV and greeks
+
+                return new OptionPriceModelResult((decimal)npv,
+                            tryGetImpliedVol,
+                            () => new Greeks(evalDeltaGamma,
+                                            () => tryGetGreekOrReevaluate(() => option.vega(), reevalVega),
+                                            () => tryGetGreekOrReevaluate(() => option.theta(), reevalTheta),
+                                            () => tryGetGreekOrReevaluate(() => option.rho(), reevalRho),
+                                            () => tryGetGreek(() => option.elasticity())));
```

Comment:
This is not a final solution, and I thought about allowing user to extend this functionality at the moment we offer him to extend pricing models. So this will be refactored before becoming public. Just putting closures now into static methods will pull a lot of captured vars and signatures will not be meaningful.